src/draw/mod.rs

```diff
@@ -212,6 +227,10 @@ mod inner {
         fn clear(&mut self);
         /// Translate this context by (dx, dy) from its current position
         fn translate(&self, dx: f64, dy: f64);
+        /// Gets the current absolute x offset.
+        fn get_x_offset(&self) -> f64;
```

Comment:
Why would a widget need to know it's position on the screen? It should be self contained

Reply:
It's similar to the way you're drawing the bar in the first place, you start with some absolute position and increment it by the width allocated to each Widget.  The only difference is that when you're doing it for the bar you know you can start at zero, but when I'm doing it for a window list I can't assume that.

My consumer code is https://github.com/dunmatt/thecat/blob/47a2c535926c3b3cd9a40fa862ef18d76758a192/src/widgets/window_list.rs#L114 , perhaps you can think of some way to do the same thing without this method?

Reply:
Huh...that's actually quite surprising :confused: You _should_ be able to just use the `translate` method on the context to bump it along depending on that offset you are computing rather than setting the context value absolutely like that. Have you tried that? Something like:

```rust
        for text in text_box_iter!(self) {
            text.draw(ctx, screen, screen_has_focus, greedy_width, h)?;
            let dx = if text.is_greedy() {
                greedy_width
            } else {
               text.current_extent(ctx, h)?.0
            };
            ctx.translate(dx, 0.0);
        }
```

Reply:
Wouldn't relative translations fail in cases where  the contained widgets also [call `translate`](https://github.com/sminez/penrose/blob/588c7771a31681d215e551859b84be9f0a4d08bd/src/draw/bar/widgets.rs#L295)?

Reply:
aren't you only using text widgets? Those shouldn't be doing any context translation themselves...

If you are using other widgets as well then this sounds like something that should be a more generic "widget that lays out other widgets" thing and it would be good to understand what it is that you feel is missing from the current functionality.

Reply:
I'm only using text widgets at the moment, but the plan is to add interesting new functionality (eg per-window resource usage graphs) there once I get the basics going.  But that's a bit beside the point, the more general point is that the correctness of the container shouldn't depend on the correctness of the contained.  Both because it's a violation of the idea of containment, and because long range dependencies like that build up into rats nests over time.

Wasn't your original idea that all widgets could contain other widgets?  I thought (but I can't check right now) that one of the comments in the Text widget was something to the effect of "for use as a component of other widgets"?  I'd be happy to help work through a layout widget with you, but that may be out of scope for this PR?

Reply:
Yeah, text widgets are intended as shared logic for handling rendering text, and widgets shouldn't have to care about their container. That's why I'm surprised that translating like I've suggested doesn't work. if you look at the redraw method on the status bar, the X offset is set absolutely after rendering each widget. So any intervening translations shouldn't affect the status bar as a whole.
Your wrapper widget should be able to translate with relative offsets based on the starting position. You shouldn't need to query and set using absolute coordinates...unless I'm missing something?

Reply:
I did look at the status bar, that's why I wrote my code the way I did (why think if you can borrow inspiration from upstream?).  If you take seriously the idea that widgets shouldn't have to care about their container, then it is absolutely essential that their container uses absolute position, because otherwise the widgets would have to care about their containers' expectations.

I think there are two things you're missing, one is the point that the correctness of containers shouldn't depend on the correctness of their contents, and the other may be that the container I'm writing calls draw on multiple sub-widgets per redrawing (with its absolute repositioning) of the status bar.